pallets/liquidity-pools/src/tests/inbound.rs

```diff
@@ -0,0 +1,792 @@
+		fn with_wrong_pool() {
+			System::externalities().execute_with(|| {
+				config_mocks();
+				Pools::mock_pool_exists(|_| false);
+
+				assert_noop!(
+					LiquidityPools::submit(
+						EVM_DOMAIN_ADDRESS,
+						Message::IncreaseRedeemOrder {
+							pool_id: POOL_ID,
+							tranche_id: TRANCHE_ID,
+							investor: ALICE.into(),
+							currency: util::currency_index(CURRENCY_ID),
+							amount: AMOUNT,
+						},
+					),
+					Error::<Runtime>::PoolNotFound,
+				);
+			})
+		}
+
+		#[test]
+		fn with_wrong_tranche() {
+			System::externalities().execute_with(|| {
+				config_mocks();
+				Pools::mock_tranche_exists(|_, _| false);
+
+				assert_noop!(
+					LiquidityPools::submit(
+						EVM_DOMAIN_ADDRESS,
+						Message::IncreaseRedeemOrder {
+							pool_id: POOL_ID,
+							tranche_id: TRANCHE_ID,
+							investor: ALICE.into(),
+							currency: util::currency_index(CURRENCY_ID),
+							amount: AMOUNT,
+						},
+					),
+					Error::<Runtime>::TrancheNotFound,
+				);
+			})
+		}
+
+		#[test]
+		fn with_no_metadata() {
+			System::externalities().execute_with(|| {
+				config_mocks();
+				AssetRegistry::mock_metadata(|_| None);
+
+				assert_noop!(
+					LiquidityPools::submit(
+						EVM_DOMAIN_ADDRESS,
+						Message::IncreaseRedeemOrder {
+							pool_id: POOL_ID,
+							tranche_id: TRANCHE_ID,
+							investor: ALICE.into(),
+							currency: util::currency_index(CURRENCY_ID),
+							amount: AMOUNT,
+						},
+					),
+					Error::<Runtime>::AssetNotFound,
+				);
+			})
+		}
+
+		#[test]
+		fn without_sufficient_balance() {
+			System::externalities().execute_with(|| {
+				config_mocks();
+
+				assert_noop!(
+					LiquidityPools::submit(
+						EVM_DOMAIN_ADDRESS,
+						Message::IncreaseRedeemOrder {
+							pool_id: POOL_ID,
+							tranche_id: TRANCHE_ID,
+							investor: ALICE.into(),
+							currency: util::currency_index(CURRENCY_ID),
+							amount: AMOUNT,
+						},
+					),
+					orml_tokens::Error::<Runtime>::BalanceTooLow
+				);
+			})
+		}
+	}
+}
+
+mod handle_cancel_redeem_order {
+	use super::*;
+
+	fn config_mocks() {
+		DomainAccountToDomainAddress::mock_convert(|_| CENTRIFUGE_DOMAIN_ADDRESS);
+		DomainAddressToAccountId::mock_convert(|_| ALICE);
+		ForeignInvestment::mock_redemption(|_, _| Ok(AMOUNT));
+		Pools::mock_pool_exists(|_| true);
+		Pools::mock_tranche_exists(|_, _| true);
+		AssetRegistry::mock_metadata(|_| Some(util::default_metadata()));
+		ForeignInvestment::mock_decrease_foreign_redemption(
+			|who, investment_id, amount, foreign_currency| {
+				assert_eq!(*who, ALICE);
+				assert_eq!(investment_id, INVESTMENT_ID);
+				assert_eq!(amount, AMOUNT);
+				assert_eq!(foreign_currency, CURRENCY_ID);
+
+				// Side effects of this call
+				ForeignInvestment::mock_redemption(|_, _| Ok(0));
+				Tokens::mint_into(TRANCHE_CURRENCY, &ALICE, AMOUNT).unwrap();
+				Ok(())
+			},
+		);
+		Gateway::mock_submit(|sender, destination, msg| {
+			assert_eq!(sender, TreasuryAccount::get());
```

Comment:
Double-check, should it not be the local representation of the receiver? similar to what is used in `TransferTrancheTokens` message: https://github.com/centrifuge/centrifuge-chain/blob/main/pallets/liquidity-pools/src/inbound.rs#L95

Reply:
I think that works as well but it is not necessary because assets are not supposed to be transferred in `n-hops`, e.g. from one EVM domain to another with Centrifuge as middleman. However, we support those hops for tranche token transfers.